docs/proposals/multipart_trasnfer_mode.md

```diff
@@ -0,0 +1,296 @@
+```
+
+## Expected HTTP Responses
+
+For each one of the `parts`, as well as `abort` and `verify` requests sent by the client, the following responses are
+to be expected:
+
+* Any response with a `20x` status code is to be considered by clients as successful. This ambiguity is by design, to
+support variances between vendors (which may use `200` or `201` to indicate a successful upload, for example).
+
+* Any other response is to be considered as an error, and it is up to the client to decide whether the request should
+be retried or not. Implementors are encouraged to follow standard HTTP error status code guidelines.
+
+### `batch` replies for partially uploaded content
+When content was already partially uploaded, the server is expected to return a normal reply but omit request and parts
+which do not need to be repeated. If the entire file has been uploaded, it is expected that no `actions` value will be
+returned, in which case clients should simply skip the upload.
+
+However, if parts of the file were successfully uploaded while others weren't, it is expected that a normal reply would
+be returned, but with less `parts` to send.
+
+### `verify` HTTP 409 errors
+An `HTTP 409` error on `verify` requests typically indicates that the file could not be fully committed or verified.
+In this  case, clients should follow the following process to try and recover from the error:
+* retry the `batch` request to see if any parts of the file were not uploaded yet. If there are still `parts` to upload
+  (i.e. `parts` is not empty), proceed to upload them and re-do `verify`
+* If `parts` is empty, it is possible that the file exists in storage but is corrupt / has wrong size. In this case
+  it is recommended to issue an `abort` and re-attempt the same upload again
+* It is recommended to take special note of the number of retries, to avoid infinite recovery attempt loops
+
+## Additional Considerations
+
+### Chunk sizing
+It is up to the LFS server to decide the size of each file chunk.
+
+### Action lifetime considerations
+As multipart uploads tend to require much more time than simple uploads, it is recommended to allow for longer `expires_in`
+values than one would consider for `basic` uploads. It is possible that the process of uploading a single object in multiple
+parts may take several hours from `batch` to `verify`.
+
+### Falling back to `basic` transfer for small files
+Using multipart upload APIs has some complexity and speed overhead, and for this reason it is recommended that servers
+implement a "fallback" to `basic` transfers if the uploaded object is small enough to handle in a single part.
+
+Clients *should* support such fallback natively, as it rides on existing transfer method negotiation capabilities.
+
+The server can simply respond with `{"transfer": "basic", ...}`, even if `mutipart` was request by the client
+and *is supported* by the server in order to achieve this.
```

Comment:
Typo: “mutipart” for “multipart”.  Also, let's not make the transfers magical, so let's permit servers to fall back if both `basic` and `multipart` were specified.  That makes it a lot easier for the client to have a little more control over things, and theoretically have a simple, multipart-only implementation if they know the server they use supports that.

Reply:
Yes, that was generally my intention. The assumption was that clients should continue to always support and specify `basic`, but I'm ok with dropping that assumption and saying something like "if the client specified both `multipart` and `basic` as requested transfer modes, the server *may* decide to fall back to `basic` transfer if ..."

Reply:
Typo fixed